CSC8503CoreClasses/PhysicsSystem.cpp

```diff
@@ -220,51 +221,138 @@ In tutorial 5, we start determining the correct response to a collision,
-	Vector3 inertiaB = Vector3::Cross(physB->GetInertiaTensor() * Vector3::Cross(relativeB, p.normal), relativeB);
-	float angularEffect = Vector3::Dot(inertiaA + inertiaB, p.normal);
+	float angularEffect = CalculateInertia(physA, physB,  relativeA,  relativeB,  p.normal);
 
 	float cRestitution = 0.66f; // loss of kinetic energy
 
 	float j = (-(1.0f + cRestitution) * impulseForce) / (totalMass + angularEffect);
-
-	Vector3 fullImpulse = p.normal * j;
+	Vector3 fullImpulse = p.normal * j * physA->GetElasticity();
 
 	// apply impulse in opposite directions for collision responce
 	physA->ApplyLinearImpulse(-fullImpulse);
 	physB->ApplyLinearImpulse(fullImpulse);
 	physA->ApplyAngularImpulse(Vector3::Cross(relativeA, -fullImpulse));
 	physB->ApplyAngularImpulse(Vector3::Cross(relativeB, fullImpulse));
+
+	return j;
+}
+
+void PhysicsSystem::FrictionImpulse(GameObject& a, GameObject& b, CollisionDetection::ContactPoint& p, float j) const {
+	PhysicsObject* physA = a.GetPhysicsObject();
+	PhysicsObject* physB = b.GetPhysicsObject();
+	float totalMass = physA->GetInverseMass() + physB->GetInverseMass();
+	Vector3 relativeA = p.localA;
+	Vector3 relativeB = p.localB;
+
+	Vector3 contactVelocity;
+
+	contactVelocity = (GetIsCapsule(a) || GetIsCapsule(b) ? contactVelocity = physB->GetLinearVelocity() - physA->GetLinearVelocity() : contactVelocity = CalculateCollisionVelocity(relativeA, relativeB, physA, physB));
+
+	Vector3 tangent = CalculateFrictionDirection(contactVelocity, p.normal);
+	float angularEffect = CalculateInertia(physA, physB, relativeA, relativeB, tangent);
+
+	float impulseForce = Vector3::Dot(contactVelocity, tangent);
+	
+	j = abs(j);
+
+	float friction = CalculateFriction(physA, physB);
+
+	Vector3 fullFrictionImpulse = CalculateFrictionImpulse(tangent, j, friction);
+
+	// apply impulse in opposite directions for collision responce
+	physA->ApplyLinearImpulse(fullFrictionImpulse);
+	physB->ApplyLinearImpulse(-fullFrictionImpulse);
+	physA->ApplyAngularImpulse(Vector3::Cross(relativeA, fullFrictionImpulse));
+	physB->ApplyAngularImpulse(Vector3::Cross(relativeB, -fullFrictionImpulse));
+}
+
+bool PhysicsSystem::GetIsCapsule(GameObject& obj) const {
+	if (obj.GetBoundingVolume()->type == VolumeType::Capsule)
+		return true;
+	return false;
+}
+
+void PhysicsSystem::SeperateObjects(GameObject& a, GameObject& b, CollisionDetection::ContactPoint& p, float totalMass) const {
+	Transform& transformA = a.GetTransform();
+	Transform& transformB = b.GetTransform();
+
+	transformA.SetPosition(transformA.GetPosition() - (p.normal * p.penetration * (a.GetPhysicsObject()->GetInverseMass() / totalMass)));
+	transformB.SetPosition(transformB.GetPosition() + (p.normal * p.penetration * (b.GetPhysicsObject()->GetInverseMass() / totalMass)));
+}
+
+bool PhysicsSystem::CheckFrictionShuldBeApplied(float aVelocity, float bVelocity, float totalMass) const {
+	if (aVelocity < 5.5 && bVelocity < 5.5)
+		return false;
+	if (totalMass == 0)
```

Comment:
Is there any way mass could (presumably incorrectly) be less than 0? Do we need to check for it?